components/file_system/src/rate_limiter.rs

```diff
@@ -1,148 +1,558 @@
+#[derive(Debug)]
+struct RawIORateLimiter {
+    // IO amount passed through within current epoch
+    ios_through: [CachePadded<AtomicUsize>; IOPriority::COUNT],
+    // Maximum IOs permitted within current epoch
+    ios_per_epoch: [CachePadded<AtomicUsize>; IOPriority::COUNT],
+    protected: Mutex<RawIORateLimiterProtected>,
+}
+
+#[derive(Debug)]
+struct RawIORateLimiterProtected {
+    next_refill_time: Instant,
+    // IO amount that is drew from the next epoch in advance
+    pending_ios: [usize; IOPriority::COUNT],
+    // estimated throughput of recent epochs
+    estimated_ios_through: [IOThroughputEstimator; IOPriority::COUNT],
+}
+
+impl RawIORateLimiterProtected {
+    fn new() -> Self {
+        RawIORateLimiterProtected {
+            next_refill_time: Instant::now_coarse() + DEFAULT_REFILL_PERIOD,
+            pending_ios: [0; IOPriority::COUNT],
+            estimated_ios_through: [IOThroughputEstimator::new(); IOPriority::COUNT],
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct IOThroughputEstimator {
+    /// total count of sampled epochs
+    count: usize,
+    /// sum of IOs amount over a window
+    sum: usize,
+}
+
+impl IOThroughputEstimator {
+    fn new() -> IOThroughputEstimator {
+        IOThroughputEstimator { count: 0, sum: 0 }
+    }
+
+    fn maybe_update_estimation(&mut self, v: usize) -> Option<usize> {
+        const WINDOW_SIZE: usize = 5;
+        self.count += 1;
+        self.sum += v;
+        // average over DEFAULT_REFILL_PERIOD*WINDOW_SIZE=200ms window
+        if self.count % WINDOW_SIZE == 0 {
+            let avg = self.sum / WINDOW_SIZE;
+            self.sum = 0;
+            Some(avg)
+        } else {
+            None
+        }
+    }
+}
+
+/// Actual implementation for requesting IOs from RawIORateLimiter.
+/// An attempt will be recorded first. If the attempted amount exceeds the available quotas of
+/// current epoch, the requester will register itself and sleep until next epoch.
+macro_rules! request_imp {
+    ($self:ident, $priority:ident, $amount:ident, $mode:tt) => {{
+        let priority_idx = $priority as usize;
+        loop {
+            let cached_ios_per_refill = $self.ios_per_epoch[priority_idx].load(Ordering::Relaxed);
+            if cached_ios_per_refill == 0 {
+                return $amount;
+            }
+            let amount = std::cmp::min($amount, cached_ios_per_refill);
+            let ios_through =
+                $self.ios_through[priority_idx].fetch_add(amount, Ordering::AcqRel) + amount;
+            if ios_through <= cached_ios_per_refill {
+                return amount;
+            }
+            let now = Instant::now_coarse();
+            let (next_refill_time, pending) = {
+                let mut locked = $self.protected.lock();
+                // a small delay in case a refill slips in after `ios_per_epoch` was fetched.
+                if locked.next_refill_time + Duration::from_millis(1) >= now + DEFAULT_REFILL_PERIOD
```

Comment:
In order to speed up this specific condition? Because a mutex seems necessary to make sure a `CAS`-like semantics between `refill_time` and `pending_ios`